### PR TITLE
deepin.deepin-camera: init at 1.4.8

### PR DESCRIPTION
**Pull request by wineee**
###### Description of changes
https://github.com/linuxdeepin/dde-nixos/issues/9

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by wineee** [comment]
@ofborg build deepin.deepin-camera

**Comment by wineee** [comment]
@ofborg build deepin.deepin-camera

**Comment by wineee** [comment]
@ofborg build deepin.deepin-camera

File: pkgs/desktops/deepin/apps/deepin-camera/default.nix [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+  };
+
+  patches = [ ./let_libPath_read_LD_LIBRARY_PATH.diff ];
+
+  postPatch = ''
+    substituteInPlace src/CMakeLists.txt \
+      --replace "/usr/share/libimagevisualresult/filter_cube" "${image-editor}/share/libimagevisualresult/filter_cube" \
+      --replace "/usr/include/libusb-1.0" "${lib.getDev libusb1}/include/libusb-1.0"
+    substituteInPlace src/com.deepin.Camera.service \
+      --replace "/usr/bin/qdbus" "qdbus" \
```
Comment (NickCao) [inline pkgs/desktops/deepin/apps/deepin-camera/default.nix:41]:
```suggestion
      --replace "/usr/bin/qdbus" "${lib.getBin qttools}/bin/qdbus" \
```

File: pkgs/desktops/deepin/apps/deepin-camera/default.nix [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+    image-editor
+    qtbase
+    qtmultimedia
+    ffmpeg
+    ffmpegthumbnailer
+    libusb1
+    portaudio
+    libv4l
+  ] ++ (with gst_all_1 ; [
+    gstreamer.dev
```
Comment (NickCao) [inline pkgs/desktops/deepin/apps/deepin-camera/default.nix:64]:
```suggestion
    gstreamer
```

File: pkgs/desktops/deepin/apps/deepin-camera/default.nix [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+
+  NIX_CFLAGS_COMPILE = [
+    "-I${gst_all_1.gstreamer.dev}/include/gstreamer-1.0"
+    "-I${gst_all_1.gst-plugins-base.dev}/include/gstreamer-1.0"
+  ];
+
+  # qt5integration must be placed before qtsvg in QT_PLUGIN_PATH
+  qtWrapperArgs = [
+    "--prefix QT_PLUGIN_PATH : ${qt5integration}/${qtbase.qtPluginPrefix}"
+    "--prefix LD_LIBRARY_PATH : ${lib.makeLibraryPath [ ffmpeg ffmpegthumbnailer gst_all_1.gstreamer gst_all_1.gst-plugins-base libusb1 libv4l portaudio (lib.getLib systemd) ]}"
```
Comment (NickCao) [inline pkgs/desktops/deepin/apps/deepin-camera/default.nix:80]: `makeLibraryPath` calls `getLib` automatically.

File: pkgs/desktops/deepin/apps/deepin-camera/default.nix [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+  version = "1.4.8";
+
+  src = fetchFromGitHub {
+    owner = "linuxdeepin";
+    repo = pname;
+    rev = version;
+    sha256 = "sha256-p2RCetx1lgLonXZaC3umE+nDgZnp64o3iR2MgQhbisM=";
+  };
+
+  patches = [ ./let_libPath_read_LD_LIBRARY_PATH.diff ];
```
Comment (NickCao) [inline pkgs/desktops/deepin/apps/deepin-camera/default.nix:34]: Ideally we should reuse the functionalities of the dynamic linker, e.g. `dlopen`. Can you provide more context about the use of `libPath` (like its call site) to see if a better solution is viable.

Comment (wineee) [reply pkgs/desktops/deepin/apps/deepin-camera/default.nix:34]: 1. Used for QLibrary,  but I don't think it's necessary，QLibrary can handle more cases than libPath 

https://github.com/linuxdeepin/deepin-camera/blob/3d3d13818a750964f2d871146f466f21d82fde85/src/src/gstvideowriter.cpp#L75

2. Get the absolute path of the dynamic library，
https://github.com/linuxdeepin/deepin-camera/blob/3d3d13818a750964f2d871146f466f21d82fde85/src/src/mainwindow.cpp#L839-L862
https://github.com/linuxdeepin/deepin-camera/blob/3d3d13818a750964f2d871146f466f21d82fde85/src/src/basepub/load_libs.c#L603

Comment (NickCao) [reply pkgs/desktops/deepin/apps/deepin-camera/default.nix:34]: So can't we just keep the lib names as is in `CMainWindow::initDynamicLibPath()`, without resolving to absolute paths?

**Comment by wineee** [comment]
@ofborg build deepin.deepin-camera

File: pkgs/desktops/deepin/apps/deepin-camera/default.nix (hunk 4 above)
Comment (wineee) [reply pkgs/desktops/deepin/apps/deepin-camera/default.nix]: Seems good enough for `dlopen` and 'QLibrary'

**Comment by wineee** [comment]
@ofborg build deepin.deepin-camera

**Review by NickCao — APPROVED** [review]

**Pull request merged**